src/librustc_codegen_ssa/mir/debuginfo.rs

```diff
@@ -104,37 +104,52 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
     // FIXME(eddyb) use `llvm.dbg.value` (which would work for operands),
     // not just `llvm.dbg.declare` (which requires `alloca`).
     pub fn debug_introduce_local(&self, bx: &mut Bx, local: mir::Local) {
-        let upvar_debuginfo = &self.mir.__upvar_debuginfo_codegen_only_do_not_use;
-
         // FIXME(eddyb) maybe name the return place as `_0` or `return`?
         if local == mir::RETURN_PLACE {
             return;
         }
 
-        let decl = &self.mir.local_decls[local];
-        let (name, kind) = if self.mir.local_kind(local) == mir::LocalKind::Arg {
+        let vars = match &self.per_local_var_debug_info {
+            Some(per_local) => &per_local[local],
+            None => return,
+        };
+        let whole_local_var = vars.iter().find(|var| {
+            var.place.projection.is_empty()
+        });
+        let has_proj = || vars.iter().find(|var| {
+            !var.place.projection.is_empty()
+        }).is_some();
+
+        let (fallback_var, kind) = if self.mir.local_kind(local) == mir::LocalKind::Arg {
             let arg_index = local.index() - 1;
 
             // Add debuginfo even to unnamed arguments.
             // FIXME(eddyb) is this really needed?
-            let name = if arg_index == 0 && !upvar_debuginfo.is_empty() {
+            let var = if arg_index == 0 && has_proj() {
                 // Hide closure environments from debuginfo.
                 // FIXME(eddyb) shouldn't `ArgumentVariable` indices
                 // be offset to account for the hidden environment?
                 None
             } else {
-                Some(decl.name.unwrap_or(kw::Invalid))
+                Some(VarDebugInfo {
+                    name: kw::Invalid,
+                    source_info: self.mir.local_decls[local].source_info,
+                    place: mir::Place {
```

Comment:
local.into() should work, too

Reply:
I kept looking for something like that and I can't find it. Can you point me to where it would be implemented?